File: coworld/tools/select_assets.py
"""Write reviewed asset lists, including external textures referenced by GLBs."""
import json
from pathlib import Path
import re
import struct

ROOT = Path(__file__).resolve().parents[2]
DATA = ROOT.parent / 'polyworld_data'
GAMES = {'gota': 'gods_of_the_arena', 'lvd': 'light_vs_dark', 'cta': 'call_to_adventure'}


def strings(source):
    """Extract string literals from a small Nim asset table."""
    return re.findall(r'"([^"\n]+)"', source)


def references(asset):
    """Read relative image and buffer URIs from a GLB's JSON chunk."""
    if asset.suffix != '.glb':
        return []
    data = asset.read_bytes()
    length, kind = struct.unpack_from('<II', data, 12)
    assert kind == 0x4e4f534a
    root = json.loads(data[20:20 + length])
    return [asset.parent / item['uri'] for key in ('images', 'buffers')
            for item in root.get(key, []) if item.get('uri') and not item['uri'].startswith('data:')]


terrain = (ROOT / 'src/polyworld/quadterrain.nim').read_text()
surfaces = strings((ROOT / 'src/polyworld/terrainsurfaces.nim').read_text().split('SurfaceNames* = [')[1].split(']')[0])
for game, directory in GAMES.items():
    graphics = (ROOT / 'examples' / directory / 'graphics.nim').read_text()
    content = (ROOT / 'examples' / directory / 'content.nim').read_text()
    selected = set()
    def add(name):
        """Include one required file or a directory loaded wholesale by the UI."""
        path = DATA / name
        assert path.exists(), name
        if path.is_dir():
            for child in path.rglob('*'):
                if child.is_file() and not child.name.startswith('.'):
                    selected.add(child.relative_to(DATA).as_posix())
        else:
            selected.add(name)
    for name in ['icons', 'themes/main', 'ui', 'fonts/Rubik-Regular.ttf',
                 'fonts/Rubik-Bold.ttf', 'fonts/OverpassMono-Regular.ttf',
                 f'themes/{game}/{game}_logo.png', 'terrain/low_poly_grass.glb']:
        add(name)
    for name in strings(terrain.split('TreeTextures = [')[1].split(']')[0]):
        add(f'terrain/handpainted_trees/{name}.png')
    tree_names = ['tree_fir_01', 'tree_fir_02']
    if game == 'cta':
        tree_names += ['tree_fir_03', 'tree_leafy_simple', 'tree_leafy_double']
    for name in tree_names:
        add(f'terrain/handpainted_trees/{name}.glb')
    for name in ['water_1_normal', 'water_2_normal']:
        add(f'terrain/water_normals/{name}.jpg')
    if game == 'cta':
        add('terrain/low_poly_rocks.glb')
        for name in strings(terrain.split('TerrainMaterials = [')[1].split(']')[0]):
            for channel in ['color', 'height']:
                add(f'terrain/cartoon_textures/{name}_{channel}.png')
    else:
        add('terrain/toon_enchanted_meadow/rocks.glb')
        for name in surfaces:
            for folder in ['tiles', 'stamps']:
                for channel in ['rgb', 'height']:
                    add(f'terrain/{folder}/{name}.{channel}.png')
        if game == 'gota':
            for name in ['mossy-building-stone-1', 'dry-stacked-stone-1']:
                for channel in ['rgb', 'height']:
                    add(f'terrain/tiles/{name}.{channel}.png')
    for name in re.findall(r'DataRoot\s*&\s*"/([^"\n]+)"', graphics + content):
        if Path(name).suffix:
            add(name)
    if game == 'gota':
        for name in re.findall(r'icon:\s*"([^"]+)"', content):
            if (DATA / 'abilities' / f'{name}.png').exists():
                add(f'abilities/{name}.png')
            elif (DATA / 'items' / f'{name}.png').exists():
                add(f'items/{name}.png')
    if game == 'cta':
        icons = content.split('AbilityIconFiles*')[1].split('] = [')[1].split(']')[0]
        for name in strings(icons):
            if name:
                add(f'abilities/{name}.png')
    if game == 'lvd':
        for name in list(selected):
            if name.startswith('characters/') and name.endswith('.glb'):
                add(str(Path(name).with_suffix('.profile.png')))
        props = graphics.split('BuildingProps = [')[1].split('BuildingPropHeights')[0]
        for name in strings(props) + ['mineral1']:
            if not name:
                continue
            for pack in ['low_poly_village', 'tower_defense_kit']:
                path = f'terrain/{pack}.{name}.profile.png'
                if (DATA / path).exists():
                    add(path)
    for name in list(selected):
        for path in references(DATA / name):
            add(path.resolve().relative_to(DATA.resolve()).as_posix())
    output = ROOT / 'examples' / directory / 'webdata.txt'
    output.write_text('# Required replay assets at the revision in coworld/assets.json.\n' + '\n'.join(sorted(selected)) + '\n')
    print(game, len(selected), round(sum((DATA / p).stat().st_size for p in selected) / 1024**2, 1), 'MiB')

File: coworld/tools/select_assets.nim
import
  std/[algorithm, json, os, sequtils, sets, strutils],
  jsony,
  tooling

proc strings(source: string): seq[string] =
  ## Extracts quoted filenames from the game's small Nim asset tables.
  var position = 0
  while position < source.len:
    let first = source.find('"', position)
    if first < 0:
      break
    let last = source.find('"', first + 1)
    require(last >= 0, "Unterminated asset filename")
    let value = source[first + 1 ..< last]
    if value.len > 0 and '\n' notin value:
      result.add value
    position = last + 1

proc assignedStrings(source, marker: string): seq[string] =
  ## Extracts string values immediately following a source expression.
  var position = 0
  while position < source.len:
    let found = source.find(marker, position)
    if found < 0:
      break
    position = found + marker.len
    while position < source.len and source[position] in Whitespace:
      inc position
    if marker == "DataRoot":
      if position >= source.len or source[position] != '&':
        continue
      inc position
      while position < source.len and source[position] in Whitespace:
        inc position
    if position < source.len and source[position] == '"':
      let last = source.find('"', position + 1)
      require(last >= 0, "Unterminated filename after " & marker)
      result.add source[position + 1 ..< last]
      position = last + 1

proc readUint32(data: string, offset: int): uint32 =
  ## Reads a checked little-endian GLB header field.
  require(offset >= 0 and offset <= data.len - 4, "Truncated GLB header")
  for i in 0 ..< 4:
    result = result or (uint32(data[offset + i].ord) shl (i * 8))

proc references(asset: string): seq[string] =
  ## Reads external image and buffer paths from a GLB's JSON chunk.
  if asset.splitFile.ext != ".glb":
    return
  let
    data = readFile(asset)
    length = data.readUint32(12).int
    kind = data.readUint32(16)
  require(kind == 0x4e4f534a'u32 and length <= data.len - 20,
    "Invalid GLB JSON chunk: " & asset)
  let document = data[20 ..< 20 + length].fromJson(JsonNode)
  for key in ["images", "buffers"]:
    if not document.hasKey(key):
      continue
    for item in document[key]:
      if item.hasKey("uri"):
        let uri = item["uri"].getStr()
        if uri.len > 0 and not uri.startsWith("data:"):
          result.add asset.parentDir / uri

proc selectAssets(data: string) =
  ## Writes each game's sorted asset list including referenced GLB textures.
  let
    terrain = readFile(Root / "src/polyworld/quadterrain.nim")
    surfaces = strings(between(
      readFile(Root / "src/polyworld/terrainsurfaces.nim"),
      "SurfaceNames* = [", "]"
    ))
  for game in Games:
    let
      source = Root / "examples" / game.directory
      graphics = readFile(source / "graphics.nim")
      content = readFile(source / "content.nim")
    var selected = initHashSet[string]()
    proc add(name: string) =
      ## Includes an asset or all nonhidden files in a required directory.
      let path = data / name
      require(fileExists(path) or dirExists(path), "Missing asset: " & path)
      if dirExists(path):
        for child in walkDirRec(path, yieldFilter = {pcFile, pcLinkToFile}):
          if not child.extractFilename().startsWith("."):
            selected.incl relativePath(child, data).replace('\\', '/')
      else:
        selected.incl name
    for name in [
      "icons", "themes/main", "ui", "fonts/Rubik-Regular.ttf",
      "fonts/Rubik-Bold.ttf", "fonts/OverpassMono-Regular.ttf",
      "themes/" & game.name & "/" & game.name & "_logo.png",
      "terrain/low_poly_grass.glb"
    ]:
      add(name)
    for name in strings(between(terrain, "TreeTextures = [", "]")):
      add("terrain/handpainted_trees/" & name & ".png")
    var trees = @["tree_fir_01", "tree_fir_02"]
    if game.name == "cta":
      trees.add ["tree_fir_03", "tree_leafy_simple", "tree_leafy_double"]
    for name in trees:
      add("terrain/handpainted_trees/" & name & ".glb")
    for name in ["water_1_normal", "water_2_normal"]:
      add("terrain/water_normals/" & name & ".jpg")
    if game.name == "cta":
      add("terrain/low_poly_rocks.glb")
      for name in strings(between(terrain, "TerrainMaterials = [", "]")):
        for channel in ["color", "height"]:
          add("terrain/cartoon_textures/" & name & "_" & channel & ".png")
    else:
      add("terrain/toon_enchanted_meadow/rocks.glb")
      for name in surfaces:
        for folder in ["tiles", "stamps"]:
          for channel in ["rgb", "height"]:
            add("terrain/" & folder & "/" & name & "." & channel & ".png")
      if game.name == "gota":
        for name in ["mossy-building-stone-1", "dry-stacked-stone-1"]:
          for channel in ["rgb", "height"]:
            add("terrain/tiles/" & name & "." & channel & ".png")
    for name in assignedStrings(graphics & content, "DataRoot"):
      if name.startsWith("/") and name.splitFile.ext.len > 0:
        add(name[1 .. ^1])
    if game.name == "gota":
      for name in assignedStrings(content, "icon:"):
        if fileExists(data / "abilities" / (name & ".png")):
          add("abilities/" & name & ".png")
        elif fileExists(data / "items" / (name & ".png")):
          add("items/" & name & ".png")
    if game.name == "cta":
      let start = content.find("AbilityIconFiles*")
      require(start >= 0, "Missing CTA ability icon table")
      for name in strings(between(content[start .. ^1], "] = [", "]")):
        add("abilities/" & name & ".png")
    if game.name == "lvd":
      for name in toSeq(selected):
        if name.startsWith("characters/") and name.endsWith(".glb"):
          add(name.changeFileExt(".profile.png"))
      let props = strings(between(
        graphics, "BuildingProps = [", "BuildingPropHeights"
      ))
      for name in props & @["mineral1"]:
        for pack in ["low_poly_village", "tower_defense_kit"]:
          let path = "terrain/" & pack & "." & name & ".profile.png"
          if fileExists(data / path):
            add(path)
    for name in toSeq(selected):
      for path in references(data / name):
        let relative = relativePath(expandFilename(path), expandFilename(data))
        require(not relative.startsWith(".." & DirSep) and relative != "..",
          "GLB reference escapes asset directory: " & path)
        add(relative.replace('\\', '/'))
    let names = sorted(toSeq(selected))
    writeFile(
      source / "webdata.txt",
      "# Required replay assets at the revision in coworld/assets.json.\n" &
        names.join("\n") & "\n"
    )
    var bytes = 0'i64
    for name in names:
      bytes += getFileSize(data / name)
    echo game.name, " ", names.len, " ",
      formatFloat(bytes.float / (1024 * 1024), ffDecimal, 1), " MiB"

proc main() =
  ## Selects assets from the configured or sibling asset repository.
  require(paramCount() == 0, "Usage: nim r coworld/tools/select_assets.nim")
  selectAssets(getEnv("POLYWORLD_DATA", Root.parentDir / "polyworld_data"))

runTool(main)
